src/utils.rs

```diff
@@ -1,15 +1,70 @@
 //! Helper routines.
 
-use std::ops::Range;
-
+use anyhow::{Error, anyhow};
 use camino::Utf8Path;
 use github_actions_models::common::{
     Env,
     expr::{ExplicitExpr, LoE},
 };
+use jsonschema::{
+    BasicOutput::{Invalid, Valid},
+    Validator,
+    output::{ErrorDescription, OutputUnit},
+    validator_for,
+};
+use std::{collections::VecDeque, ops::Range};
+use std::{fmt::Write, sync::LazyLock};
 
 use crate::audit::AuditInput;
 
+static ACTION_VALIDATOR: LazyLock<Validator> = LazyLock::new(|| {
+    validator_for(&serde_json::from_str(include_str!("../github-action.json")).unwrap()).unwrap()
+});
+
+static WORKFLOW_VALIDATOR: LazyLock<Validator> = LazyLock::new(|| {
+    validator_for(&serde_json::from_str(include_str!("../github-workflow.json")).unwrap()).unwrap()
+});
+
+pub(crate) fn validate_action(contents: String) -> Error {
+    match serde_yaml::from_str(&contents) {
+        Ok(workflow) => match ACTION_VALIDATOR.apply(&workflow).basic() {
+            Valid(_) => anyhow!("valid action but failed unmarshaling"),
```

Comment:
Not really sure how to handle this case since in theory this should be unreachable (?), maybe you have some suggestion on what to do or what error message to print out here?

Reply:
`unreachable` makes sense to me here 🙂 

```suggestion
            Valid(_) => unreachable!("valid action but invalid YAML"),
```

Reply:
This has the side effect of making this a panic, however, so the tests below should probably become integration snapshots instead -- see <https://woodruffw.github.io/zizmor/development/#writing-snapshot-tests> and <https://github.com/woodruffw/zizmor/blob/ee6f160a304b96fbde2b6fe346e71218c1f07bf0/tests/integration/e2e.rs#L184-L195> for some docs/examples 🙂

Reply:
I think maybe the tests can remain as unit tests by attempting to capture the panic instead, what do you think? For example:
https://github.com/woodruffw/zizmor/blob/f642eaf19a95534c42bf68169156eaf995a19261/src/utils.rs#L387-L389
Either way I did have to update a snapshot there so thanks for reminding me 🙂
Also, regarding your suggestion `"valid action but invalid YAML"`, this unreachable snippet is only reached when there is a discrepancy between the schema validation and the unmarshaling logic (which in theory shouldn't happen), I'm not sure if the YAML is guaranteed to be invalid in those cases? 🤔

Reply:
> Also, regarding your suggestion "valid action but invalid YAML", this unreachable snippet is only reached when there is a discrepancy between the schema validation and the unmarshaling logic (which in theory shouldn't happen), I'm not sure if the YAML is guaranteed to be invalid in those cases? 🤔

Oh yeah, I see what you mean (and I misunderstood the context for this error) -- yeah, this should probably instead indicate to the user that the bug is on `zizmor`'s modeling side, rather than an invalid workflow/action definition per the schema. Sorry for the confusion!

(This also undoes the need for the panic handling, so sorry for the wild goose chase there.)